### ISSUE TITLE
Expand vocabulary in `evaluate`

### ISSUE THREAD

**Issue by matt-gardner** [allenai/allennlp#327]
We should have a good way to expand a vocabulary in `evaluate` and similar commands, so that you can add words from a set of pre-trained vectors, or assign unique random vectors to new OOV words.  @OyvindTafjord has some preliminary code for doing this; see here:  https://github.com/allenai/allennlp/issues/239#issuecomment-329804862.  There are some sticky issues to figure out around how to tell the `evaluate` command which vectors to use, but that actual expand logic looks pretty good in Oyvind's diff.

Maybe we need to allow for an evaluate json file, that specifies a model archive and other parameters...?

**Comment by roys174** [comment]
This feature is super important for out-of-domain evaluation, and would probably also be beneficial for the online demo, so it would be great to add it. @matt-gardner 's suggestion (adding another parameter for evaluation) sounds like it could do the job.

**Comment by matt-gardner** [comment]
Just an implementation note - #590 made it so it should be relatively easy to do this without any new evaluation json file.  We just need to add some stuff to the training config (and the embedding class, and the trainer...) that adds the word embedding file to the model archive, and augments the vocabulary on load.

**Comment by schmmd** [comment]
This is also important for the `fine-tune` command.

**Comment by matthew-z** [comment]
I wonder what I can do for dataset like SQuAD that test set is hidden at this moment. Is there any way to load the whole Vocabulary from Glove instead of choosing tokens only appeared in train and dev?

**Comment by matt-gardner** [comment]
There's no easy way to do this right now, unfortunately.  PRs welcome!

**Comment by matthew-z** [comment]
Thanks! I just figured out that it is possible load vocabulary from embedding https://github.com/allenai/allennlp/pull/478/files

**Comment by matt-gardner** [comment]
From the parameter documentation: "Words which appear in the pretrained embedding file but not in the data are NOT included in the Vocabulary."

**Comment by matthew-z** [comment]
Oh.....Thank you for the reminder.... 


If these words are not included, then why we want to provide the embedding file? Both of them sound like unusual cases:

> 1. Restricting the Vocabulary to only words present in the pretrained embedding file but I feed this sounds like a very rare case.
> 2. Ensure that words are added if they are present in the pretrained embedding file, regardless of their count.

I wonder if it is better to be possible to add tokens from the embedding regardless of if the word in the count.  it seems that it is not difficult to do.

**Comment by matt-gardner** [comment]
Those uses are mostly targeted for cases where you're using _fixed_ glove embeddings (or similar), so you can use anything in the embedding file regardless of how much training data you have for it (and maybe you don't want to bother including embeddings for anything outside of glove, because the embeddings wouldn't get trained).  Including the entirely vocabulary would typically be massive, so we don't do that.

One option you have here is to create your own subclass of `Vocabulary` and have it load the whole embedding file.  This should work just fine.

**Comment by OyvindTafjord** [comment]
See https://github.com/allenai/allennlp/pull/1822 for a way to preload a large chunk (or all) of a pretrained embeddings file (example config here: https://github.com/allenai/allennlp/blob/master/training_config/quarel_parser_zeroshot_elmo.json)

**Comment by matthew-z** [comment]
Thank you!!  I think it is what I am looking for.

**Comment by matt-gardner** [comment]
Ah, thanks @OyvindTafjord, looks like the docstring for that parameter never got updated to reflect the new option.

**Comment by ramithp** [comment]
Sorry to comment on an old thread, but right now, if I were to train on one vocabulary, but intend to predict/evaluate on another domain/vocabulary, is there a way for me to get my model to use my fine-tuned embeddings for words common between the two domains and default to GloVe (and then random if GloVe is unavailable) for OOV terms in the new domain?

The example by @OyvindTafjord above did not clarify it for me. :/

**Comment by HarshTrivedi** [comment]
You can use `--extend-vocab` option in `evaluate` command (#2501).

**Comment by ramithp** [comment]
That's a neat feature! Does this apply for the predict command as well?
I need to decode translations as well.

**Comment by HarshTrivedi** [comment]
`--extend-vocab` option isn't available in `predict` (for now at least). To have a similar behavior, you would need to do 2 things using newer instances. (1) Expand the vocabulary (2) Expand the model's embedder, as in [here](https://github.com/allenai/allennlp/blob/master/allennlp/commands/evaluate.py#L150-L153). One way to do this is to use those 2 api's in your predictor itself. Eg. `predict_batch_json` can be overrided as follows.

Change

```python
    def predict_batch_json(self, inputs: List[JsonDict]) -> List[JsonDict]:
        instances = self._batch_json_to_instances(inputs)
        return self.predict_batch_instance(instances)
```
to

```python
    def predict_batch_json(self, inputs: List[JsonDict]) -> List[JsonDict]:
        instances = self._batch_json_to_instances(inputs)
        self._model.vocab.extend_from_instances(Params({}), instances=instances)
        self._model.extend_embedder_vocab(self._embedding_sources)
        # self._embedding_sources can be loaded in Predictor's attribute via additional
        # argument of --embedding-sources-mapping in predict command. (Same as evaluate)
        return self.predict_batch_instance(instances)
```
.. and so on.

**However,** when you have pretrained-embedding file to extend the model embedding from, the embedding extension happens by reading entire file again and plucking only those weights that are new ([here](https://github.com/allenai/allennlp/blob/master/allennlp/modules/token_embedders/embedding.py#L233-L234)). And predict command operates lazily (not all instances are available upfront). Therefore, you would end-up reading full pretrained-embedding for each new batch/instance that have new words!

But I think, with some more work you can get-away reading the entire file each time. If you manage to get it to work, it would be good to have PR for this. I was planning to take this, but haven't got time to get back to it yet.

**Comment by matt-gardner** [comment]
This is another argument for having `predict` and `evaluate` behave the same, just with different outputs. We could then just read all the data up front, like `evaluate` does, and expand the vocab. Doing this in a demo is still problematic, but we can worry about that later.

**Comment by HarshTrivedi** [comment]
+1 to that! That's seems a better long term solution!

**Comment by HarshTrivedi** [comment]
@ramithp if you are still looking for vocab expansion in `predict`, I have a `predict-with-vocab-expansion` command [here](https://github.com/StonyBrookNLP/multee/blob/master/lib/commands/predict_with_vocab_expansion.py).

**Comment by stevenwernercs** [comment]
> @ramithp if you are still looking for vocab expansion in `predict`, I have a `predict-with-vocab-expansion` command [here](https://github.com/StonyBrookNLP/multee/blob/master/lib/commands/predict_with_vocab_expansion.py).

any examples on how to call predict-with-vocab-expansion? 
Having trouble finding examples of the EMBEDDING_SOURCES_MAPPING json format expected

**Comment by HarshTrivedi** [comment]
@stevenwernercs [Here](https://github.com/StonyBrookNLP/multee#multirc) is an example.
